packages/blocks/src/store/selectors.js

```diff
@@ -109,19 +109,19 @@ export function getActiveBlockVariation( state, blockName, attributes, scope ) {
 	const variations = getBlockVariations( state, blockName, scope );
 
 	const match = variations?.find( ( variation ) => {
-		if ( typeof variation.isActive === 'function' ) {
-			return variation.isActive( attributes, variation.attributes );
-		}
-
 		if ( Array.isArray( variation.isActive ) ) {
-			return variation.isActive.every(
-				( attribute ) =>
-					attributes[ attribute ] ===
-					variation.attributes[ attribute ]
-			);
+			const blockType = getBlockType( state, blockName );
+			const attributeKeys = Object.keys( blockType.attributes || {} );
+			return variation.isActive
+				.filter( ( attribute ) => attributeKeys.includes( attribute ) )
+				.every(
+					( attribute ) =>
+						attributes[ attribute ] ===
+						variation.attributes[ attribute ]
```

Comment:
I think we'll still want to return false if either of these values is undefined. Blocks and variations may have optional attributes.

Reply:
I can't come up with a case for this 🤔 It seems to be working as expected:

1. If both of them are `undefined`, then that's correct and should match.
2. If `variation.attributes[ attribute ] = undefined` and `attributes[ attribute ] != undefined` then it won't match
3. If `variation.attributes[ attribute ] != undefined` and `attributes[ attribute ] = undefined` then it won't match

Reply:
@gwwar can you expand on the use case you have in mind with an example?

Reply:
Sorry, I missed 👀 that we were already iterating over variations instead of all possible block attributes, so only the left side may be undefined, eg attribute[ attribute ]. Code here is fine 👍 

btw I personally find that code reuse in tests aren't that useful. What's nicer for me is being able to read a single test case without needing to skip to different definitions. (Fixtures/snapshots can also help when things go crazy). Curious if folks had thoughts on that.

Here's a free test case, where I verified this: 

```js
it( 'should handle optional attributes', () => {
				const testLinkName = 'test/link-block';
				const state = {
					blockTypes: {
						[ testLinkName ]: {
							name: testLinkName,
							attributes: {
								url: { type: 'string' },
								type: { type: 'string' },
								id: { type: 'number' },
							},
						},
					},
					blockVariations: {
						[ testLinkName ]: [
							{
								attributes: { url: 'https://wordpress.org' },
								isActive: [ 'url' ],
							},
							{
								attributes: { type: 'foo' },
								isActive: [ 'type' ],
							},
							{
								attributes: {
									type: 'bar',
									url: 'https://example.com',
								},
								isActive: [ 'type', 'url' ],
							},
						],
					},
				};

				expect(
					getActiveBlockVariation( state, testLinkName, {
						type: 'foo',
					} )
				).toEqual( {
					attributes: { type: 'foo' },
					isActive: [ 'type' ],
				} );

				expect(
					getActiveBlockVariation( state, testLinkName, {
						url: 'https://wordpress.org',
					} )
				).toEqual( {
					attributes: { url: 'https://wordpress.org' },
					isActive: [ 'url' ],
				} );

				expect(
					getActiveBlockVariation( state, testLinkName, {
						type: 'bar',
						url: 'https://example.com',
					} )
				).toEqual( {
					attributes: {
						type: 'bar',
						url: 'https://example.com',
					},
					isActive: [ 'type', 'url' ],
				} );

				expect(
					getActiveBlockVariation( state, testLinkName, {
						id: 1234,
					} )
				).toEqual( undefined );
			} );
```

Reply:
Yeah, I had that in mind too. 😄 That's why the last tests are totally separate. Definitely needs some cleaning. A little bit of reuse doesn't hurt though :D

Reply:
> A little bit of reuse doesn't hurt though

Certainly, I wanted to leave a note since I was having a bit of trouble reading what the state shape was without inspecting tests.